File: config_loader.js
#!/usr/bin/env node
/**
 * Configuration Loader (JavaScript/Node.js version)
 * Demonstrates how to load region configuration from environment variables and config files.
 */

const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');

class ConfigLoader {
  /**
   * Load and manage application configuration with region settings.
   * @param {string} configFile - Path to the configuration file
   */
  constructor(configFile = 'config.yaml') {
    this.configFile = configFile;
    this.config = this._loadConfig();
  }

  /**
   * Load configuration from YAML file.
   * @private
   */
  _loadConfig() {
    try {
      const filePath = path.resolve(this.configFile);
      const fileContents = fs.readFileSync(filePath, 'utf8');
      return yaml.load(fileContents);
    } catch (e) {
      throw new Error(`Failed to load configuration file: ${e.message}`);
    }
  }

  /**
   * Get the configured region.
   * Priority: Environment variable > Config file
   * @returns {string} The configured region
   */
  getRegion() {
    // Check environment variable first
    const envRegion = process.env.AWS_REGION;
    if (envRegion) {
      console.log(`Using region from environment variable: ${envRegion}`);
      return envRegion;
    }

    // Fall back to config file
    const configRegion = this.config?.region?.primary || 'us-east-1';
    console.log(`Using region from config file: ${configRegion}`);
    return configRegion;
  }

  /**
   * Get detailed information about a region.
   * @param {string|null} region - Region to get info for (defaults to current region)
   * @returns {object} Region information
   */
  getRegionInfo(region = null) {
    if (!region) {
      region = this.getRegion();
    }

    const regionSettings = this.config?.region?.settings || {};
    return regionSettings[region] || {};
  }

  /**
   * List all available regions from config.
   * @returns {string[]} Array of available region codes
   */
  listAvailableRegions() {
    const regionSettings = this.config?.region?.settings || {};
    return Object.entries(regionSettings)
      .filter(([_, info]) => info.enabled)
      .map(([region, _]) => region);
  }

  /**
   * Get the application environment.
   * @returns {string} The current environment
   */
  getEnvironment() {
    return process.env.ENVIRONMENT || 'development';
  }
}

/**
 * Example usage of ConfigLoader.
 */
function main() {
  console.log('='.repeat(60));
  console.log('Region Configuration Loader');
  console.log('='.repeat(60));

  try {
    const config = new ConfigLoader();

    // Display current region
    const currentRegion = config.getRegion();
    console.log(`\n✓ Current Region: ${currentRegion}`);

    // Display region details
    const regionInfo = config.getRegionInfo();
    if (Object.keys(regionInfo).length > 0) {
      console.log(`  Name: ${regionInfo.name || 'N/A'}`);
      console.log(`  Availability Zones: ${regionInfo.availability_zones || 'N/A'}`);
    }

    // Display environment
    console.log(`\n✓ Environment: ${config.getEnvironment()}`);

    // List available regions
    console.log('\n✓ Available Regions:');
    config.listAvailableRegions().forEach(region => {
      const info = config.getRegionInfo(region);
      console.log(`  - ${region}: ${info.name || 'N/A'}`);
    });

    console.log('\n' + '='.repeat(60));
    console.log('To change the region, set the AWS_REGION environment variable:');
    console.log('  export AWS_REGION=eu-west-1');
    console.log('  node config_loader.js');
    console.log('='.repeat(60));

    return 0;
  } catch (e) {
    console.error(`\n✗ Error: ${e.message}`);
    return 1;
  }
}

// Run if called directly
if (require.main === module) {
  process.exit(main());
}

module.exports = ConfigLoader;

File: config_loader.py
#!/usr/bin/env python3
"""
Configuration Loader
Demonstrates how to load region configuration from environment variables and config files.
"""

import os
import yaml
from pathlib import Path


class ConfigLoader:
    """Load and manage application configuration with region settings."""
    
    def __init__(self, config_file='config.yaml'):
        self.config_file = Path(config_file)
        self.config = self._load_config()
        
    def _load_config(self):
        """Load configuration from YAML file."""
        if not self.config_file.exists():
            raise FileNotFoundError(f"Configuration file not found: {self.config_file}")
            
        with open(self.config_file, 'r') as f:
            return yaml.safe_load(f)
    
    def get_region(self):
        """
        Get the configured region.
        Priority: Environment variable > Config file
        """
        # Check environment variable first
        env_region = os.getenv('AWS_REGION')
        if env_region:
            print(f"Using region from environment variable: {env_region}")
            return env_region
        
        # Fall back to config file
        config_region = self.config.get('region', {}).get('primary', 'us-east-1')
        print(f"Using region from config file: {config_region}")
        return config_region
    
    def get_region_info(self, region=None):
        """Get detailed information about a region."""
        if region is None:
            region = self.get_region()
        
        region_settings = self.config.get('region', {}).get('settings', {})
        return region_settings.get(region, {})
    
    def list_available_regions(self):
        """List all available regions from config."""
        region_settings = self.config.get('region', {}).get('settings', {})
        return [
            region for region, info in region_settings.items()
            if info.get('enabled', False)
        ]
    
    def get_environment(self):
        """Get the application environment."""
        return os.getenv('ENVIRONMENT', 'development')


def main():
    """Example usage of ConfigLoader."""
    print("=" * 60)
    print("Region Configuration Loader")
    print("=" * 60)
    
    try:
        config = ConfigLoader()
        
        # Display current region
        current_region = config.get_region()
        print(f"\n✓ Current Region: {current_region}")
        
        # Display region details
        region_info = config.get_region_info()
        if region_info:
            print(f"  Name: {region_info.get('name', 'N/A')}")
            print(f"  Availability Zones: {region_info.get('availability_zones', 'N/A')}")
        
        # Display environment
        print(f"\n✓ Environment: {config.get_environment()}")
        
        # List available regions
        print(f"\n✓ Available Regions:")
        for region in config.list_available_regions():
            info = config.get_region_info(region)
            print(f"  - {region}: {info.get('name', 'N/A')}")
        
        print("\n" + "=" * 60)
        print("To change the region, set the AWS_REGION environment variable:")
        print("  export AWS_REGION=eu-west-1")
        print("  python config_loader.py")
        print("=" * 60)
        
    except Exception as e:
        print(f"\n✗ Error: {e}")
        return 1
    
    return 0


if __name__ == '__main__':
    exit(main())
